README.md

<div style="width: 100%;height: 300px;">
    <img src="./.github/image.jpg" style="width: 100%;object-fit:cover">
</div>


<div align="center">
    <hr/>
    <a href="https://github.com/soft-cute/test">
        <img src="https://img.shields.io/badge/github-soft--cute%2Ftest-1707320?logo=github">
    </a>
    <img src="https://img.shields.io/github/last-commit/soft-cute/test/master?logo=github">
    <a href="https://github.com/soft-cute/test/releases">
        <img src="https://img.shields.io/github/v/release/soft-cute/test?logo=github">
    </a>
    <img src="https://img.shields.io/github/release-date/soft-cute/test">
    <img src="https://img.shields.io/github/license/soft-cute/test">
</div><br><br>

# 目录
- [目录](#目录)
- [说明](#说明)
    - [命令行使用方法](#命令行使用方法)
    - [是否开启录制或播放](#是否开启录制或播放)
    - [使用配置文件](#使用配置文件)
- [注意](#注意)

# 说明

```bash
____________________________________________________

                        _
          ___  __   __ (_)  _ __     __ _
         / _ \ \ \ / / | | | '_ \   / _` |
        |  __/  \ V /  | | | | | | | (_| |
         \___|   \_/   |_| |_| |_|  \__,_|

____________________________________________________


Usage: evina.exe [OPTIONS]
       evina.exe [OPTIONS] <COMMAND>

Commands:
  config  编辑配置文件
  help    Print this message or the help of the given subcommand(s)

Options:
  -l, --live <LIVE>  需要录制的平台 [possible values: douyu, douyin]
  -i, --id <ID>      直播间号或链接
  -x <X>             播放器的宽度 [default: 1366]
  -y <Y>             播放器的高度 [default: 768]
  -d <DOWNLOAD>      自定义录制的目录，默认保存到根目录下的download [default: /download]
  -m, --ffm          是否开启录播，默认不开启
  -p, --ffp          是否开启播放功能，默认不开启
  -r, --read         使用配置文件配置的id
  -f, --file <FILE>  自定义配置文件的路径 [default: C:\Users\softcute\.evina\config]
  -h, --help         Print help
  -V, --version      Print version

```



### 命令行使用方法
```shell
# 斗鱼平台仅支持输入房间ID
evina --live douyu --id 123456
或者
evina -l douyu -i 123456

# 抖音平台支持输入链接和房间ID
# 如果使用手机分享的短链接，必须配置抖音账号的 cookie
evina --live douyin --id https://v.douyin.com/abcdefg/
evina --live douyin --id abcdefg/
evina --live douyin --id abcdefg
evina --live douyin --id https://live.douyin.com/123456789
evina --live douyin --id 123456789
或者
evina -l douyin -i https://v.douyin.com/abcdefg/
evina -l douyin -i abcdefg/
evina -l douyin -i abcdefg
evina -l douyin -i https://live.douyin.com/123456789
evina -l douyin -i 123456789
```
### 是否开启录制或播放
```shell
# 开启录制功能
evina -l douyu -i 123456 -m
# 开启录制功能并设置保存目录
# 目录格式默认为 自定义目录/直播间名字/日期时间/录制的视频
evina -l douyu -i 123456 -m -d 自定义目录
# 开启播放功能
evina -l douyu -i 123456 -p
# 开启播放功能并设置分辨率大小
evina -l douyu -i 123456 -p -x 1080 -y 768
# 同时开启录制和播放功能
evina -l douyu -i 123456 -mp
```
### 使用配置文件
```shell
# 创建配置文件或者格式化
evina config -r
# 自定义配置文件的路径
evina -f /aa/aa.txt config -r
# 添加参数到文件
evina config -a DOUYU_URL__1=123456
# 删除文件的参数
evina config -d DOUYU_URL__1
# 创建配置文件的软连接
evina config -s aaa/config
# 列出配置文件所有的参数
evina config -l
```

# 注意
 1. 录播需要安装FFMPEG
 2. 在线播放需要安装FFPLAY




.github/CHANGELOG.md
# Changelog



## [0.2.39] (2023-01-22)
修改：
- 斗鱼平台rtmp原画修改为蓝光4M，原画录制的文件太大
- 修改录制目录日期格式：%Y年%m月%d日-%H时%M分%S秒 --> %Y-%m-%d
- 简化了部分代码

## [0.0.1] (2023-01-22)
这是v0.0.1的测试

## [0.0.2] (2023-10-01)
这是v0.0.2的测试

[0.0.1]: https://github.com/soft-cute/test/releases/tag/v0.0.1

cli/src/main.rs

use clap::Parser;
use live::{Cli, Config};
mod live;

fn main() {
    let cli = live::Cli::parse();
    match cli.live.clone() {
        Some(data) => {
            if data == "douyu" {
                live::Information::to_print(&live::douyu::douyu(cli.id.unwrap()).unwrap());
            } else if data == "douyin" {
                live::Information::to_print(&live::douyin::douyin(cli.id.unwrap()).unwrap());
            }
        }
        None => {
            if cli.read {
                let list = vec!["DOUYU","DOUYIN"];
                live::thread_run(cookie::read_config(cli.file.clone(), list));

            }
            else {
                subcommand(cli)
            }
        }
    }
}

fn subcommand(cli: Cli) {
    match cli.sub {
        Some(Config::Config { reload, add, del, list, symlink }) => {
            if reload == true {
                cookie::live::reload(cli.file.clone());
            }else if list == true {
                cookie::live::list(cli.file.clone())
                
            }
            match add {
                Some(data) => cookie::live::add(cli.file, data),
                None => match del {
                    Some(data) => cookie::live::del(cli.file, data),
                    None => match symlink {
                        Some(data) => cookie::live::symlink(cli.file,data),
                        None => std::process::exit(0),
                    },
                },
            }
        }
        None => std::process::exit(0),
    }
}



cookie/src/lib.rs
pub mod live;
use std::collections::HashMap;
pub fn local_cookie(path:String,key: String) -> Result<String, std::io::Error> {
    let cookie = live::douyin::local_cookie(path,&key);
    return cookie;
}

pub async fn auto_cookie(url: String) -> Result<String, Box<dyn std::error::Error>> {
    let cookie = live::douyin::auto_cookie(url).await;
    return cookie;
}

pub fn read_config(path:String,prefix_list: Vec<&str>) -> HashMap<String, Option<String>> {
    let mut map = HashMap::new();
    for platform in prefix_list {
        let hashmap = dotenv_rs::get_vars_with_prefix(path.clone(), platform);
        match hashmap {
            Ok(hashmap) => map.extend(hashmap),
            Err(err) => {
                println!("{}",err.to_string());
                std::process::exit(0)
            },
        }
    };
    return map;
}


#[cfg(test)]
mod tests {
    use std::env;

    #[test]
    fn main() {
        // 获取命令行参数
        let args: Vec<String> = env::args().collect();

        // 打印程序名称
        let program_name = &args[0];
        println!("Program name: {}", program_name);

        // 打印所有命令行参数
        println!("Command line arguments:");
        for arg in args.iter().skip(1) {
            println!("{}", arg);
        }

        // 检查命令行参数数量
        if args.len() < 3 {
            println!("Usage: {} <arg1> <arg2>", program_name);
        } else {
            // 获取命令行参数
            let arg1 = &args[1];
            let arg2 = &args[2];

            // 在这里处理命令行参数
            println!("Argument 1: {}", arg1);
            println!("Argument 2: {}", arg2);
        }
    }
}


.github/check/check.sh
#!/bin/bash

workflow_runs=$(curl -H "Accept: application/vnd.github.v3+json" -H "Authorization: token $INPUT_TOKEN" https://api.github.com/repos/$INPUT_OWNER/$INPUT_REPO/actions/runs?status=in_progress)
num=$(echo $workflow_runs | jq -r .total_count)
if [ "$num" -eq 0 ]; then
  echo "in_progress=false" >> $GITHUB_OUTPUT
else
    for ((i=0; i<$num; i++)); do
        run_number=$(echo $workflow_runs | jq -r .workflow_runs[$i].run_number)
        name=$(echo $workflow_runs | jq -r .workflow_runs[$i].name)
        jobs_url=$(echo $workflow_runs | jq -r .workflow_runs[$i].jobs_url)
        echo $run_number
        echo $INPUT_RUN_NUMBER
        if [[ "$name" == "$INPUT_NAME" && "$run_number" != "$INPUT_RUN_NUMBER" ]]; then
            jobs=$(curl -H "Accept: application/vnd.github.v3+json" -H "Authorization: token $INPUT_TOKEN" $jobs_url)
            num_job=$(echo $jobs | jq -r .total_count)
            for ((i_job=0; i_job<$num_job; i_job++)); do
                name_job=$(echo $jobs | jq -r .jobs[$i_job].name)
                if [[ "$name_job" == "$INPUT_NAME_JOB" ]]; then
                    steps=$(echo $jobs | jq -r .jobs[$i_job].steps[$(($INPUT_NUM_STEP - 1))])
                    name_step=$(echo $steps | jq -r .name)
                    if [[ "$name_step" == "$INPUT_NAME_STEP" ]]; then
                        if [[ "$(echo $steps | jq -r .status)" != "completed" ]]; then
                            echo $steps | jq -r .status
                            echo "in_progress=true" >> $GITHUB_OUTPUT
                        else
                            echo "in_progress=false" >> $GITHUB_OUTPUT
                        fi
                    fi
                fi
            done
        else
            echo "in_progress=false" >> $GITHUB_OUTPUT
        fi
    done
fi

cli/src/live/mod.rs
pub mod douyin;
pub mod douyu;
use boa_engine::Context;
use chrono::Local;
use clap::{Parser, Subcommand};
use fake::locales::*;
use fake::Fake;
use regex::Regex;
use reqwest;
use std::{process::exit, process::Command, thread};

use self::douyin::douyin;
use self::douyu::douyu;

#[derive(Parser)]
#[command(author = env!("CARGO_PKG_NAME"),version = env!("CARGO_PKG_VERSION"),
about=r"
____________________________________________________

                        _                 
          ___  __   __ (_)  _ __     __ _ 
         / _ \ \ \ / / | | | '_ \   / _` |
        |  __/  \ V /  | | | | | | | (_| |
         \___|   \_/   |_| |_| |_|  \__,_|

____________________________________________________
                                   ",
long_about = None,)]
#[command(subcommand_negates_reqs = true)]
#[command(arg_required_else_help = true)]
pub struct Cli {
    /// 需要录制的平台
    #[arg(short = 'l', long,value_parser=["douyu", "douyin"])]
    pub live: Option<String>,
    /// 直播间号或链接
    #[arg(short = 'i', long)]
    pub id: Option<String>,
    /// 播放器的宽度
    #[arg(short = 'x',default_value_t = String::from("1366"))]
    pub x: String,
    /// 播放器的高度
    #[arg(short = 'y',default_value_t = String::from("768"))]
    pub y: String,
    /// 自定义录制的目录，默认保存到根目录下的download
    #[arg(short = 'd',default_value_t = String::from("/download"))]
    pub download: String,
    /// 是否开启录播，默认不开启
    #[arg(short = 'm', long)]
    pub ffm: bool,
    /// 是否开启播放功能，默认不开启
    #[arg(short = 'p', long)]
    pub ffp: bool,
    /// 使用配置文件配置的id
    #[arg(short = 'r', long)]
    pub read: bool,
    #[cfg(target_os = "windows")]
    /// 自定义配置文件的路径
    #[arg(short='f',long,default_value_t={
    let root = std::env::var("USERPROFILE").ok().unwrap();
    format!("{}\\.evina\\config",root)
})]
    pub file: String,
    #[cfg(any(target_os = "linux", target_os = "macos"))]
    /// 自定义配置文件的路径
    #[arg(short='f',long,default_value_t={
    let root = std::env::var("HOME").ok().unwrap();
    format!("{}/.evina/config",root)
})]
    pub file: String,
    #[command(subcommand)]
    pub sub: Option<Config>,
}

#[derive(Debug, Subcommand)]
pub enum Config {
    /// 编辑配置文件
    Config {
        /// 格式化配置文件
        #[arg(short = 'r', long)]
        reload: bool,
        /// 列出配置文件的所有配置
        #[arg(short = 'l', long)]
        list: bool,
        /// 添加变量到配置文件
        #[arg(short = 'a', long)]
        add: Option<String>,
        /// 删除配置文件的变量
        #[arg(short = 'd', long)]
        del: Option<String>,
        /// 创建配置文件的软链接
        #[arg(short = 's', long)]
        symlink: Option<String>,
    },
}

#[derive(Debug)]
pub struct Information {
    platform: String,
    id: String,
    name: String,
    rtmp: String,
}

impl Information {
    pub fn to_print(&self) {
        let name = if self.platform == "斗鱼" {
            "斗鱼主播名: ".to_string() + &self.name
        } else {
            "抖音主播名: ".to_string() + &self.name
        };
        println!("{}{} | 房间号ID: {}", "\n", name, self.id);
        println!("{}{}", "\n", self.rtmp);
        if !self.rtmp.contains("未开播") {
            Information::ff(self);
        }
    }

    fn ff(&self) {
        let cli = Cli::parse();
        let mut tasks = Vec::new();

        let ffm: Option<Result<thread::JoinHandle<()>, std::io::Error>> = if cli.ffm == true {
            Some(Information::ffm(self))
        } else {
            None
        };

        let ffp: Option<Result<thread::JoinHandle<()>, std::io::Error>> = if cli.ffp == true {
            Some(Information::ffp(self))
        } else {
            None
        };

        tasks.push(ffm);
        tasks.push(ffp);
        for task in tasks {
            let _ = match task {
                Some(t) => t.unwrap().join(),
                None => Ok(()),
            };
        }
    }

    fn ffm(&self) -> Result<thread::JoinHandle<()>, std::io::Error> {
        let cli = Cli::parse();
        // let fmt = "%Y年%m月%d日-%H时%M分%S秒";
        let fmt = "%Y-%m-%d";
        let now = Local::now().format(fmt);
        let path = std::path::Path::new(&cli.download).join(format!("{}录播/{}/{now}", self.platform,self.name));
        let save = format!("{}/%Y-%m-%d-%H-%M-%S.mp4", path.display());
        let _ = std::fs::create_dir_all(path);

        let ffmpeg = format!(
            r#"ffmpeg -t 19800 -i "{}" -c:a copy -c:v copy -f segment -segment_time 3600 -strftime 1 "{save}""#,
            self.rtmp
        );
        let ffmpeg = shell_words::split(&ffmpeg).unwrap();
        let ffm = thread::Builder::new()
            .name("ffm".to_owned())
            .spawn(move || {
                Command::new(ffmpeg[0].clone())
                .args(&ffmpeg[1..])
                .output()
                .expect("录播程序错误");
            });
        return ffm;
    }

    fn ffp(&self) -> Result<thread::JoinHandle<()>, std::io::Error> {
        let cli = Cli::parse();
        let info: Information = if self.platform == "斗鱼" {
            douyu::douyu(self.id.clone()).unwrap()
        } else if self.platform == "抖音" {
            douyin::douyin(self.id.clone()).unwrap()
        } else {
            exit(0x0100)
        };
        
        let ffplay = format!(r#"ffplay -x {} -y {} -i "{}""#, cli.x, cli.y, info.rtmp);
        let ffplay = shell_words::split(&ffplay).unwrap();

        let ffp = thread::Builder::new()
            .name("ffp".to_owned())
            .spawn(move || {
                Command::new(ffplay[0].clone())
                .args(&ffplay[1..])
                .output()
                .expect("播放程序错误");
            });
        return ffp;
    }
}

// 判断 id 是否包含字母
pub fn is_alphabetic(id: String) -> bool {
    let mut res = false;
    for c in id.chars() {
        if c.is_alphabetic() {
            res = true;
            break;
        }
    }
    return res;
}

pub fn useag() -> String {
    use fake::faker::internet::raw::*;
    loop {
        let add = UserAgent(ZH_CN).fake::<String>();
        if !add.contains("iPhone")
            && !add.contains("Windows CE")
            && !add.contains("Android")
            && !add.contains("WIndows NT")
            && !add.contains("Windows NT")
            && !add.contains("Linux i686")
            && !add.contains("iPad")
        {
            return add;
        }
    }
}

pub fn re_result(re: &str, data: String) -> Result<String, std::io::Error> {
    let result = Regex::new(re)
        .unwrap()
        .captures(&data)
        .map(|caps| caps.get(1).unwrap().as_str().to_owned());
    match result {
        Some(result) => Ok(result),
        None => {
            println!("{}", "未找到直播间");
            std::process::exit(0x0100)
        }
    }
}
pub async fn eval(js: &str) -> String {
    let mut context = Context::default();
    match context.eval(js) {
        Ok(result) => result.display().to_string(),
        Err(e) => e.display().to_string(),
    }
}

pub fn thread_run(hashmap: std::collections::HashMap<String, Option<String>>) {
    let mut tasks = Vec::new();
    for (key, value) in hashmap.iter() {
        let key = key.clone();
        let value = value.clone().unwrap();
        let run = thread::Builder::new().name(value.clone()).spawn(move || {
            if key.contains("DOUYU") && value != "xxxxxx" {
                Information::to_print(&douyu(value).unwrap())
            } else if key.contains("DOUYIN") && value != "xxxxxx" {
                Information::to_print(&douyin(value).unwrap())
            }
        });
        tasks.push(run);
    }

    for task in tasks {
        let _ = match task {
            Ok(t) => t.join(),
            Err(_) => todo!(),
        };
    }
}
